pages/mixins.py

```diff
@@ -47,28 +49,74 @@ def get_account():
 
 class UploadedImageMixin(object):
 
-    def get_media(self, storage, filter_list):
-        list_media = self.list_media(storage, filter_list)
-        if len(list_media) == 1:
-            return list_media[0]
-        return None
+    @staticmethod
+    def update_media_tag(tags, list_media):
+        for item in list_media['results']:
+            media_tags = MediaTag.objects.filter(location=item['location'])
+            new_media_tags = []
+            if tags:
+                for tag in tags:
+                    if tag != "":
+                        new_media_tag, _ = MediaTag.objects.get_or_create(
+                            tag=tag, location=item['location'])
+                        new_media_tags += [new_media_tag]
+
+            # compare new list and delete removed tags.
+            for tag in media_tags:
+                if not tag in new_media_tags:
+                    tag.delete()
+
+    @staticmethod
+    def delete_media_items(storage, list_media):
+        for item in list_media['results']:
+            storage.delete(item['media'])
+
+            # Delete all MediaTag and PageElement using this location
+            MediaTag.objects.filter(location=item['location']).delete()
+            PageElement.objects.filter(text=item['location']).delete()
+
+    @staticmethod
+    def build_filter_list(request):
+        items = request.DATA.get('items')
+
+        if not items and request.DATA.get('location'):
+            items = [{'location': request.DATA.get('location')}]
+
+        filter_list = []
+        if items:
+            for item in items:
+                filter_list = item['location']
+        return filter_list
 
     @staticmethod
-    def list_media(storage, filter_list):
```

Comment:
Create two methods ``fitler`` and ``exclude`` so that semantics match the one of Django ``QuerySet``. Both *include* and *exclude* filters are implemented in the same method using a ``delete`` parameter flag here. It is just to complicated to understand and use correctly.